faunadb-scala/src/main/scala/faunadb/values/Time.scala

```diff
@@ -0,0 +1,89 @@
+package faunadb.values
+package time
+
+import org.joda.time.Instant
+import org.joda.time.format.ISODateTimeFormat
+import scala.Ordered._
+
+/**
+  * Represents the instant from Java's epoch composed by the number of seconds since eposh
+  * and the nanoseconds offset.
+  */
+class HighPrecisionTime private(val secondsSinceEposh: Long, val nanoSecondsOffset: Int) {
+
+  /** Returns a [[org.joda.time.Instant]]. Truncates micro and nanoseconds. */
+  def toInstant: Instant =
+    new Instant(milliSecondsSinceEpoch)
+
+  /** Milliseconds since Java epoch. Truncates micro and nanoseconds. */
+  def milliSecondsSinceEpoch: Long =
+    secondsSinceEposh / 1000
+
+  override def equals(obj: Any): Boolean = obj match {
+    case other: HighPrecisionTime =>
+      secondsSinceEposh == other.secondsSinceEposh &&
+        nanoSecondsOffset == other.nanoSecondsOffset
+
+    case _ => false
+  }
+
+  override def hashCode(): Int =
+    secondsSinceEposh.hashCode() + nanoSecondsOffset.hashCode()
+
+  override def toString =
+    "%s%06dZ".format(
+      toInstant.toString(ISODateTimeFormat.dateHourMinuteSecondMillis()),
+      nanoSecondsOffset
+    )
+}
+
+object HighPrecisionTime {
+
+  private val NanosInAMicro = 10000
+  private val NanosInASecond = 1000000000
+
+  /**
+    * Creates an instance of [[faunadb.values.time.HighPrecisionTime]] adding
+    * the the micro and nanoseconds informed.
+    */
+  def apply(initial: Instant, microsToAdd: Long = 0, nanosToAdd: Long = 0): HighPrecisionTime = {
+    val nanos = microsToAdd * NanosInAMicro + nanosToAdd
+    val overflow = nanos / NanosInASecond
+    val remaining = nanos % NanosInASecond
+
+    new HighPrecisionTime(initial.getMillis * 1000 + overflow, remaining.toInt)
```

Comment:
This could probably be simplified if you extract the partial second via the regex below and pass it in as nanos, using instant only for the seconds.